### ISSUE TITLE
FreeBSD support

### ISSUE THREAD

**Issue by englishm-llnw** [saltstack/kitchen-salt#133]
The `/bin/sh` shipped with FreeBSD appears to be far more strict (and minimalist?) about adherence to IEEE Std 1003.1 ("POSIX.1") than any `sh` shipped with popular Linux distros.

I suspect this is why several errors are raised when `kitchen-salt` attempts to run on the platform:

```
$ kitchen converge freebsd
-----> Starting Kitchen (v1.16.0)
-----> Converging <default-freebsd-12>...
       Preparing files for transfer
       Preparing salt-minion
       Preparing pillars into /srv/pillar
       Preparing state collection
       neither collection_name or formula have been set, assuming this is a pr
e-built collection                                                           
       Preparing state_top
       Preparing scripts into /etc/salt/scripts
       Unmatched '''.
       #: Command not found.
       Badly placed ()'s.
       Ambiguous output redirect.
       then: Command not found.
       return: Command not found.
       else: endif not found.
>>>>>> ------Exception-------
>>>>>> Class: Kitchen::ActionFailed
>>>>>> Message: 1 actions failed.
>>>>>>     Converge failed on instance <default-freebsd-12>.  Please see .kitc
hen/logs/default-freebsd-12.log for more details                             
>>>>>> ----------------------
>>>>>> Please see .kitchen/logs/kitchen.log for more details
>>>>>> Also try running `kitchen diagnose --all` for configuration
$ 
```

Where `freebsd-12` is defined in `.kitchen.yml` as:

```yaml
driver:
  name: vagrant

platforms:
  - name: freebsd-12
    driver:
      box: freebsd/FreeBSD-12.0-CURRENT
      ssh:
        shell: '/usr/local/bin/bash'
...
```

The first error appears to be due to the newline following `sh -c '` and things go downhill quickly.

Here is where the relevant section of code begins:

https://github.com/saltstack/kitchen-salt/blob/24841701ea7880c9b39068a03275237fb1485705/lib/kitchen/provisioner/install.erb#L11-L19

And here is where the contents of `Util.shell_helpers` begin:

[test-kitchen/test-kitchen@v1.16.0 - support/download_helpers.sh:1-10](https://github.com/test-kitchen/test-kitchen/blob/v1.16.0/support/download_helpers.sh#L1-L10)

```sh
# Check whether a command exists - returns 0 if it does, 1 if it does not
exists() {
  if command -v $1 >/dev/null 2>&1
  then
    return 0
  else
    return 1
  fi
}
```

This appears to correspond to the series of errors above.

**Comment by gtmanfred** [comment]
Unfortunately I know very little about FreeBSD, so I am probably not going to be much help, but would love to see a PR to make FreeBSD supported.

Thanks,
Daniel

**Comment by gtmanfred** [comment]
Hrm, that is odd, because that error is caused by `csh`

```
$ csh -c '
> exists() {
>   if command -v $1 >/dev/null 2>&1
>   then
>     return 0
>   else
>     return 1
>   fi
> }
> exists nope
> echo $?
> '
Badly placed ()'s.
Ambiguous output redirect.
then: Command not found.
return: Command not found.
else: endif not found.
```

```
$ sh -c '
> exists() {
>   if command -v $1 >/dev/null 2>&1
>   then
>     return 0
>   else
>     return 1
>   fi
> }
> exists nope
> echo $?
> '
1
```

**Comment by gtmanfred** [comment]
Also worth noting that `csh` doesn't allow you to have multiline scripts on the commandline

```
vagrant@:~ % sh -c '
Unmatched '''.
```

If we can figure out how to let that happen, it should be able to be worked around.

**Comment by gtmanfred** [comment]
@englishm-llnw how are you able to get the freebsd box to boot?

I keep running into this error

```
No base MAC address was specified. This is required for the NAT networking
       to work properly (and hence port forwarding, SSH, etc.). Specifying this
       MAC address is typically up to the box and box maintainer. Please contact
       the relevant person to solve this issue.
```

Using this kitchen.yml

https://github.com/gtmanfred/kitchen-salt/blob/master/.kitchen.yml#L109-L112

**Comment by englishm-llnw** [comment]
@gtmanfred Thanks for looking at this!

Trying to test your branch @ a8727c82b42d2f000d52920c1a24c94f54a1741d now.

I get the same error the first time I try to create a brand-new VM with that basebox, too:

---

<details>
<summary><code>No base MAC address was specified.</code></summary>
<pre>
$ bundle exec kitchen verify freebsd
-----> Starting Kitchen (v1.17.0)
-----> Creating <bootstrap-freebsd-12>...
       Bringing machine 'default' up with 'virtualbox' provider...
       ==> default: Importing base box 'freebsd/FreeBSD-12.0-CURRENT'...
No base MAC address was specified. This is required for the NAT networking
       to work properly (and hence port forwarding, SSH, etc.). Specifying this
       MAC address is typically up to the box and box maintainer. Please contact
       the relevant person to solve this issue.
>>>>>> ------Exception-------
>>>>>> Class: Kitchen::ActionFailed
>>>>>> Message: 1 actions failed.
>>>>>>     Failed to complete #create action: [Expected process to exit with [0], but received '1'
---- Begin output of vagrant up --no-provision --provider virtualbox ----
STDOUT: Bringing machine 'default' up with 'virtualbox' provider...
==> default: Importing base box 'freebsd/FreeBSD-12.0-CURRENT'...
STDERR: No base MAC address was specified. This is required for the NAT networking
to work properly (and hence port forwarding, SSH, etc.). Specifying this
MAC address is typically up to the box and box maintainer. Please contact
the relevant person to solve this issue.
---- End output of vagrant up --no-provision --provider virtualbox ----
Ran vagrant up --no-provision --provider virtualbox returned 1] on bootstrap-freebsd-12
>>>>>> ----------------------
>>>>>> Please see .kitchen/logs/kitchen.log for more details
>>>>>> Also try running `kitchen diagnose --all` for configuration
</pre>
</details>

---

If I retry, it gets past that error and on to another (see below). We've seen this MAC address issue on occasion just using our own FreeBSD-derived basebox images, with Vagrant+VirtualBox too. I'll see if I can get whatever fix or workaround we came up with sent upstream.

The next error is from Vagrant complaining about the default shell:

---

<details>
<summary><code>The configured shell (config.ssh.shell) is invalid and unable to properly execute commands.</code></summary>
<pre>
$ bundle exec kitchen create freebsd
-----> Starting Kitchen (v1.17.0)
-----> Creating <bootstrap-freebsd-12>...
       Bringing machine 'default' up with 'virtualbox' provider...
       ==> default: Setting the name of the VM: bootstrap-freebsd-12_default_1503505123633_22816
       ==> default: Fixed port collision for 22 => 2222. Now on port 2206.
       ==> default: Clearing any previously set network interfaces...
       ==> default: Preparing network interfaces based on configuration...
           default: Adapter 1: nat
       ==> default: Forwarding ports...
           default: 22 (guest) => 2206 (host) (adapter 1)
       ==> default: Booting VM...
       ==> default: Waiting for machine to boot. This may take a few minutes...
           default: SSH address: 127.0.0.1:2206
           default: SSH username: vagrant
           default: SSH auth method: private key
           default: Warning: Remote connection disconnect. Retrying...
           default: Warning: Remote connection disconnect. Retrying...
           default: Warning: Remote connection disconnect. Retrying...
           default: Warning: Remote connection disconnect. Retrying...
           default: Warning: Remote connection disconnect. Retrying...
           default: Warning: Remote connection disconnect. Retrying...
           default: Warning: Remote connection disconnect. Retrying...
       The configured shell (config.ssh.shell) is invalid and unable
       to properly execute commands. The most common cause for this is
       using a shell that is unavailable on the system. Please verify
       you're using the full path to the shell and that the shell is
       executable by the SSH user.
>>>>>> ------Exception-------
>>>>>> Class: Kitchen::ActionFailed
>>>>>> Message: 1 actions failed.
>>>>>>     Failed to complete #create action: [Expected process to exit with [0], but received '1'
---- Begin output of vagrant up --no-provision --provider virtualbox ----
STDOUT: Bringing machine 'default' up with 'virtualbox' provider...
==> default: Setting the name of the VM: bootstrap-freebsd-12_default_1503505123633_22816
==> default: Fixed port collision for 22 => 2222. Now on port 2206.
==> default: Clearing any previously set network interfaces...
==> default: Preparing network interfaces based on configuration...
    default: Adapter 1: nat
==> default: Forwarding ports...
    default: 22 (guest) => 2206 (host) (adapter 1)
==> default: Booting VM...
==> default: Waiting for machine to boot. This may take a few minutes...
    default: SSH address: 127.0.0.1:2206
    default: SSH username: vagrant
    default: SSH auth method: private key
    default: Warning: Remote connection disconnect. Retrying...
    default: Warning: Remote connection disconnect. Retrying...
    default: Warning: Remote connection disconnect. Retrying...
    default: Warning: Remote connection disconnect. Retrying...
    default: Warning: Remote connection disconnect. Retrying...
    default: Warning: Remote connection disconnect. Retrying...
    default: Warning: Remote connection disconnect. Retrying...
STDERR: The configured shell (config.ssh.shell) is invalid and unable
to properly execute commands. The most common cause for this is
using a shell that is unavailable on the system. Please verify
you're using the full path to the shell and that the shell is
executable by the SSH user.
---- End output of vagrant up --no-provision --provider virtualbox ----
Ran vagrant up --no-provision --provider virtualbox returned 1] on bootstrap-freebsd-12
>>>>>> ----------------------
>>>>>> Please see .kitchen/logs/kitchen.log for more details
>>>>>> Also try running `kitchen diagnose --all` for configuration

$ 
</pre>
</details>

---

With the following patch, we can get past that error, too:

```diff
diff --git a/.kitchen.yml b/.kitchen.yml
index 7ed779e..55465a7 100644
--- a/.kitchen.yml
+++ b/.kitchen.yml
@@ -110,6 +110,8 @@ platforms:
     driver:
       box: freebsd/FreeBSD-12.0-CURRENT
       name: vagrant
+      ssh:
+        shell: "/usr/local/bin/bash"
 
 suites:
   - name: bootstrap
```

---

<details>
<summary><code>Finished creating &lt;bootstrap-freebsd-12&gt;</code></summary>
<pre>
$ bundle exec kitchen create freebsd
-----> Starting Kitchen (v1.17.0)
-----> Creating <bootstrap-freebsd-12>...
       Bringing machine 'default' up with 'virtualbox' provider...
       ==> default: Machine not provisioned because `--no-provision` is specified.
       [SSH] Established
       Vagrant instance <bootstrap-freebsd-12> created.
       Finished creating <bootstrap-freebsd-12> (0m4.15s).
-----> Kitchen is finished. (0m4.76s)
</pre>
</details>

---

But then we're back to errors with the install script:

---

<pre>
$ bundle exec kitchen verify freebsd
-----> Starting Kitchen (v1.17.0)
-----> Converging <bootstrap-freebsd-12>...
       Preparing files for transfer
       Preparing salt-minion
       Preparing pillars into /srv/pillar
       Preparing grains into /etc/salt/grains
       Preparing formula: tests from /Users/menglish/git/kitchen-salt
       Preparing state_top
       Preparing scripts into /etc/salt/scripts
       Preparing formula: foo from ./tests/formula-foo
       /bin/sh: Event not found.
       #: Command not found.
       Badly placed ()'s.
       Ambiguous output redirect.
       then: Command not found.
       return: Command not found.
       else: endif not found.
>>>>>> ------Exception-------
>>>>>> Class: Kitchen::ActionFailed
>>>>>> Message: 1 actions failed.
>>>>>>     Converge failed on instance <bootstrap-freebsd-12>.  Please see .kitchen/logs/bootstrap-freebsd-12.log for
 more details                                                                                                       
>>>>>> ----------------------
>>>>>> Please see .kitchen/logs/kitchen.log for more details
>>>>>> Also try running `kitchen diagnose --all` for configuration
</pre>

---

At this point, it's starting to look like Vagrant's `config.ssh.shell` setting might not be honored Kitchen's `install_command`. Need to do some more digging.

**Comment by gtmanfred** [comment]
This is the upstream bug that needs to be fixed for us to be able to use the freebsd managed vagrant boxes.

https://github.com/test-kitchen/test-kitchen/issues/683

Until that is fixed, you will have to use the bento freebsd boxes.

**Comment by gtmanfred** [comment]
yes, config.ssh.shell isn't actually used by vagrant when you do `vagrant ssh` either.  So I have labeled it as an upstream issue, until either test-kitchen or vagrant resolves this.

I am working on adding the bento/freebsd-11.0 box to the kitchen-salt tests, and once a 12.0 box is on bento, i will add it too, at least until we can use the freebsd boxes.

Thanks,
Daniel

**Comment by gtmanfred** [comment]
And looks like that is documented.

```
config.ssh.shell - The shell to use when executing SSH commands from Vagrant. By default this is bash -l. Note that this has no effect on the shell you get when you run vagrant ssh. This configuration option only affects the shell to use when executing commands internally in Vagrant.
```

https://www.vagrantup.com/docs/vagrantfile/ssh_settings.html

**Comment by englishm-llnw** [comment]
@gtmanfred Yep, that does look like the right upstream bug. Thanks!

I peeked ahead a bit by running `sudo pw usermod -n vagrant -s /bin/sh` via `kitchen login` and trying to push `verify` along a bit further along. 

It next gets stuck some bashisms and Linux-isms in [`dependencies.erb`](https://github.com/saltstack/kitchen-salt/blob/be57e99f5f854dc9e739b423bb468ad6ff286232/lib/kitchen/provisioner/dependencies.erb), beginning with the [shebang](https://github.com/saltstack/kitchen-salt/blob/be57e99f5f854dc9e739b423bb468ad6ff286232/lib/kitchen/provisioner/dependencies.erb#L103) - FreeBSD's `bash` actually lives at `/usr/local/bin/bash`, if present.

So that first error looks like:

```
       sudo: unable to execute /tmp/kitchen/dependencies.sh: No such file or directory
```

The message here isn't clear, but it's `/bin/bash` that doesn't exist.

**Comment by gtmanfred** [comment]
Yup, i am working on a fix to get the freebsd-11.0 bento box to work, i have changed /bin/bash to /usr/bin/env bash, so it should work.

Once that is in, I am going to close this issue.

Thanks!
Daniel
